go/client/install_osx.go

```diff
@@ -0,0 +1,568 @@
+	dest, err := filepath.EvalSymlinks(linkPath)
+	if err == nil && binPath != dest {
+		err = fmt.Errorf("We are not symlinked to %s", linkPath)
+	}
+	if err != nil {
+		if force {
+			return createCommandLine(binPath, linkPath)
+		}
+		return fmt.Errorf("We are not symlinked to %s", linkPath)
+	}
+
+	return nil
+}
+
+func createCommandLine(binPath string, linkPath string) error {
+	if _, err := os.Lstat(linkPath); err == nil {
+		err := os.Remove(linkPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	return os.Symlink(binPath, linkPath)
+}
+
+func installService(g *libkb.GlobalContext, force bool) error {
+	g.Log.Info("Checking service")
+	keybaseStatus := keybaseServiceStatus(g, "", "")
+	g.Log.Info("Service: %s", keybaseStatus.InstallStatus.ToString())
+	if keybaseStatus.NeedsInstall() || force {
+		g.Log.Info("Installing Keybase service")
+		uninstallKeybaseServices(g.Env.GetRunMode())
+		err := installKeybaseService(g)
+		if err != nil {
+			g.Log.Errorf("Error installing Keybase service: %s", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func installKBFS(g *libkb.GlobalContext, force bool) error {
+	g.Log.Info("Checking KBFS")
+	kbfsStatus := kbfsServiceStatus(g, "", "")
+	g.Log.Info("KBFS: %s", kbfsStatus.InstallStatus.ToString())
+	if kbfsStatus.NeedsInstall() || force {
+		g.Log.Info("Installing KBFS")
+		uninstallKBFSServices(g.Env.GetRunMode())
+		err := installKBFSService(g)
+		if err != nil {
+			g.Log.Errorf("Error installing KBFS: %s", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func binPath() (string, error) {
+	if libkb.IsBrewBuild {
+		// Use the brew opt bin directory.
+		binName := filepath.Base(os.Args[0])
+		return filepath.Join("/usr/local/opt", binName, "bin", binName), nil
+	}
+
+	path := os.Args[0]
+	if !strings.HasPrefix(path, "/") {
+		return path, fmt.Errorf("We need to be run with an absolute path to this executable to determine its full path.")
+	}
+
+	return path, nil
+}
+
+func binName() string {
+	return filepath.Base(os.Args[0])
+}
+
+func kbfsBinPath(runMode libkb.RunMode) (string, error) {
+	path, err := binPath()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(filepath.Dir(path), kbfsBinName(runMode)), nil
+}
+
+func kbfsBinName(runMode libkb.RunMode) string {
+	switch runMode {
+	case libkb.DevelRunMode:
+		return "kbfsdev"
+
+	case libkb.StagingRunMode:
+		return "kbfsstage"
+
+	case libkb.ProductionRunMode:
+		return "kbfs"
+
+	default:
+		panic("Invalid run mode")
+	}
+}
+
+func kbfsMountPath(homeDir string, runMode libkb.RunMode) string {
+	switch runMode {
+	case libkb.DevelRunMode:
+		return filepath.Join(homeDir, "Keybase.devel")
+
+	case libkb.StagingRunMode:
+		return filepath.Join(homeDir, "Keybase.staging")
+
+	case libkb.ProductionRunMode:
+		return filepath.Join(homeDir, "Keybase")
+
+	default:
+		panic("Invalid run mode")
+	}
+}
+
+func BrewAutoInstall(g *libkb.GlobalContext) error {
+	g.Log.Debug("+ AutoInstall for launchd")
```

Comment:
Added back in

Seems error prone though since you have to always do err =? (like, you could accidentally return err and the log statement would be wrong :)